[RISCV] Generate MIPS load/store pair instructions

djtodoro opened the pull request:
Introduce RISCVLoadStoreOptimizer MIR Pass that will do the optimization. The load/store pairing pass identifies adjacent load/store instructions operating on consecutive memory locations and merges them into a single paired instruction.

This is part of MIPS extensions for the p8700 CPU.

Production of ldp/sdp instructions is OFF by default, since it is beneficial for -Os only in the case of p8700 CPU.

djtodoro committed:
[RISCV] Generate MIPS load/store pair instructions

Introduce RISCVLoadStoreOptimizer MIR Pass that will do
the optimization. The load/store pairing pass identifies adjacent
load/store instructions operating on consecutive memory locations
and merges them into a single paired instruction.

This is part of MIPS extensions for the p8700 CPU.

Production of ldp/sdp instructions is OFF by default, since it
is beneficial for -Os only in the case of p8700 CPU.

topperc commented on llvm/lib/Target/RISCV/CMakeLists.txt:
```diff
@@ -48,6 +48,7 @@ add_llvm_target(RISCVCodeGen
   RISCVISelLowering.cpp
   RISCVLandingPadSetup.cpp
   RISCVMachineFunctionInfo.cpp
+  RISCVLoadStoreOptimizer.cpp
```
alphabetize

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,364 @@
+//===----------------------------------------------------------------------===//
+//
+// Bundle loads and stores that operate on consecutive memory locations to take
+// the advantage of hardware load/store bonding.
```
"the" is unnecesary i this sentence.

djtodoro committed:
fixup: Alphabetize cmake file

topperc commented:
I'm pretty sure some of this is copied from AArch64. That's a helpful thing to mention in the description as it makes it easier to review if we can cross reference the prior art. Can you mention any notable differences?

mshockwave commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,364 @@
+  }
+
+  if (!tryConvertToLdStPair(First, Second))
+    finalizeBundle(MBB, First.getInstrIterator(),
```
why do you want to bundle them when the pairing fails?

djtodoro replied:
I have introduced a new option for `bonding` that should make this more obvious now.

topperc replied:
How does that compare to load/store clustering which was already enabled in the scheduler? https://github.com/llvm/llvm-project/pull/73789

michaelmaitland replied:
Is this the same question as what was discussed here (doing it in scheduler == DAG cluster mutation): https://github.com/llvm/llvm-project/pull/117865#discussion_r1861085475?

djtodoro replied:
Bonding part is very similar thing, so I will remove this part for now and keep `ldp/sdp` optimization only in this change. I still see benefits in terms of performance when using both, `load/store clustering` from the scheduler and `bundling for bonding here`, but it can be a separate topic/change.

mshockwave commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -562,6 +570,11 @@ void RISCVPassConfig::addPreEmitPass() {
   addPass(createRISCVMakeCompressibleOptPass());
+
+  // LoadStoreOptimizer creates bundles for load-store bonding.
+  addPass(createUnpackMachineBundles([](const MachineFunction &MF) {
```
KCFI already adds UnpackMachineBundles Pass in `addPreEmitPass2`, is it possible to merge with it?

djtodoro replied:
We do not generate bundles anymore, so this is not needed.

mshockwave commented on llvm/test/CodeGen/RISCV/load-store-pair.ll:
```diff
@@ -0,0 +1,509 @@
+
+declare dso_local void @sinkf(float noundef, float noundef, float noundef, float noundef) local_unnamed_addr
+
+define dso_local void @testd(double* nocapture noundef readonly %a) local_unnamed_addr #0 {
```
nit: `testf` and `testd` seems to be negative tests (please correct me if I'm wrong), could you either add comments or rename the function to make it more obvious?

lenary commented:
Can you talk (in a comment on this PR is fine) about how you see this fitting in with e.g. Zilsd support, which we were told is coming soon to upstream from NXP?

The NXP branch against 18.0 does similar things. IIRC, it has two load-store pairing passes (an early one and a late one), it makes changes to prolog-epilog inserter, and it might also make some changes directly to isel. Maybe some of these changes are useful to Mips' pairing instructions too? The NXP changes are here: https://github.com/llvm/llvm-project/compare/release/18.x...nxp-auto-tools:llvm-project:Zilsd/release/18.1.6 (this is a compare view of their 18.1.6 version against upstream release/18.x)

I would like to see coordination between you and NXP - maybe @anmolparalkar-nxp is the right person? - so that we can get a reasonably general load/store pair codegen implementation that can support extensions where the instruction details might be different, but the overall extensions are similar.

lenary commented:
Also cc'ing @christian-herber-nxp who is involved on the Zilsd side from NXP.

dtcxzyw reviewed (commented):
Do we really need a new pass to merge two consecutive loads/stores?
XTheadMemPair codegen is supported by TargetDAGCombine:  https://reviews.llvm.org/D144002

christian-herber-nxp commented:
> Also cc'ing @christian-herber-nxp who is involved on the Zilsd side from NXP.

thanks, I will keep an eye on this. @anmolparalkar-nxp is the right person though.

djtodoro commented:
> I'm pretty sure some of this is copied from AArch64. That's a helpful thing to mention in the description as it makes it easier to review if we can cross reference the prior art. Can you mention any notable differences?

Yes. That Pass was our motivation back then, and yes, I will add that note. Thanks @topperc.

> Can you talk (in a comment on this PR is fine) about how you see this fitting in with e.g. Zilsd support, which we were told is coming soon to upstream from NXP?
> 
> The NXP branch against 18.0 does similar things. IIRC, it has two load-store pairing passes (an early one and a late one), it makes changes to prolog-epilog inserter, and it might also make some changes directly to isel. Maybe some of these changes are useful to Mips' pairing instructions too? The NXP changes are here: [release/18.x...nxp-auto-tools:llvm-project:Zilsd/release/18.1.6](https://github.com/llvm/llvm-project/compare/release/18.x...nxp-auto-tools:llvm-project:Zilsd/release/18.1.6) (this is a compare view of their 18.1.6 version against upstream release/18.x)
> 
> I would like to see coordination between you and NXP - maybe @anmolparalkar-nxp is the right person? - so that we can get a reasonably general load/store pair codegen implementation that can support extensions where the instruction details might be different, but the overall extensions are similar.

@lenary yes it makes sense, thanks! I will talk to @anmolparalkar-nxp :)

> Do we really need a new pass to merge two consecutive loads/stores? XTheadMemPair codegen is supported by TargetDAGCombine: https://reviews.llvm.org/D144002

@dtcxzyw This might be an alternative, but I am not sure the `DAGCombiner` can catch/handle more complex cases, for example this one: https://github.com/llvm/llvm-project/pull/117865#discussion_r1871131220.

anmolparalkar-nxp commented:
> @lenary yes it makes sense, thanks! I will talk to @anmolparalkar-nxp :)
Thanks @lenary. @djtodoro we will work together to streamline and unify the approach, thank you.

djtodoro committed:
fixup: Address comments

It was not very obvious that bonding and pairing are two different
things, so it is improved now. In addition, mentioned the AArch64
optimization and similarity with it.

djtodoro commented:
> I'm pretty sure some of this is copied from AArch64. That's a helpful thing to mention in the description as it makes it easier to review if we can cross reference the prior art. Can you mention any notable differences?

I have added a top-level comment that explains this. Basically, we use `tryToPairLdStInst` from it only. AArch64 performs away more things, and it can be future work for RISC-V as well.

djtodoro commented:
One part of the 32-bit extension that NXP will upstream soon can easily be extended on top of this optimization, and I do not see any conflict with this. cc @lenary @anmolparalkar-nxp

djtodoro committed:
fixup: Format and add option for scan limit

topperc commented on llvm/lib/Target/RISCV/RISCVSubtarget.cpp:
```diff
@@ -62,10 +62,15 @@ static cl::opt<unsigned> RISCVMinimumJumpTableEntries(
-                             cl::desc("RISCV: Optimize for load-store bonding"),
-                             cl::init(false), cl::Hidden);
+static cl::opt<bool> UseMIPSLoadStorePairsOpt(
+    "mips-riscv-load-store-pairs",
```
The options should start with `riscv-`. Starting with `mips-` makes it alphabetized with options for the Mips target in the help.

djtodoro replied:
Yes, that is true.

topperc commented on llvm/lib/Target/RISCV/RISCVSubtarget.cpp:
```diff
@@ -62,10 +62,15 @@ static cl::opt<unsigned> RISCVMinimumJumpTableEntries(
-                             cl::init(false), cl::Hidden);
+static cl::opt<bool> UseMIPSLoadStorePairsOpt(
+    "mips-riscv-load-store-pairs",
+    cl::desc("RISCV: Enable the load/store pair optimization pass"),
```
I don't think "RISCV" is needed in this descritpion since its already in the option name. But if was, it should always be spelled "RISC-V" per RVI branding guidelines.

djtodoro committed:
fixup: address comments

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,384 @@
+                                    bool MergeForward) {
+  MachineBasicBlock::iterator E = I->getParent()->end();
+  MachineBasicBlock::iterator NextI = next_nodbg(I, E);
+  if (NextI == Paired)
```
Please add the comment block from AArch64 for this `if`

```
  // If NextI is the second of the two instructions to be merged, we need        
  // to skip one further. Either way we merge will invalidate the iterator,      
  // and we don't need to scan the new instruction, as it's a pairwise           
  // instruction, which we're not considering for further action anyway. 
```

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,384 @@
+    return false;
+  MachineInstrBuilder MIB = BuildMI(
+      *MF,
+      First->getDebugLoc().get() ? First->getDebugLoc() : Second->getDebugLoc(),
```
Does AArch64 do this with the debug location? Looks like they always use the First unless I'm mistaken?

djtodoro commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,384 @@
+  Align MMOAlign = MMO->getAlign();
+  if (const PseudoSourceValue *Source = MMO->getPseudoValue())
+    if (Source->kind() == PseudoSourceValue::FixedStack)
+      MMOAlign = MF->getSubtarget().getFrameLowering()->getStackAlign();
```
@topperc Hmm, got it... What can be an alternative for this? Shall we check the alignment of the specific object e.g. `getObjectAlignment(Source->getFrameIndex())`?

topperc replied:
The alignment should already be correct in the `MMO`. If it's not, that's a bug in the creation of the `MMO`. We shouldn't have to check anything additional.

djtodoro replied:
@topperc Well, the idea here is, even though the alignment is `!= 16 bytes`, we rely on ABI and assume it is 16 bytes. Is it right approach?

topperc replied:
The stack pointer has to be 16 byte aligned at all times. The objects on the stack are not guaranteed to be aligned. On RV32,  4 caller saved registers are saved into a single 16 byte region. Only 1 of the 4 load/stores for those 4 registers is 16 byte aligned.

djtodoro committed:
fixup: address remaining comments

djtodoro commented on llvm/test/CodeGen/RISCV/load-store-pair.ll:
```diff
@@ -0,0 +1,510 @@
+; testd and testf look for bonding only.
+; RUN: llc -mtriple=riscv32 -verify-machineinstrs < %s \
+; RUN:   | FileCheck %s -check-prefix=RV32I
+; RUN: llc -mtriple=riscv32 -target-abi ilp32d -mattr=+d -verify-machineinstrs < %s \
```
Addressed

djtodoro commented on llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:
```diff
@@ -310,7 +311,12 @@ RISCVTargetMachine::createMachineScheduler(MachineSchedContext *C) const {
   ScheduleDAGMI *DAG = nullptr;
-  if (EnablePostMISchedLoadStoreClustering) {
+  const RISCVSubtarget &ST = C->MF->getSubtarget<RISCVSubtarget>();
+  bool EnableLoadStoreClusteringForLoadStoreOpt =
```
Well, it is tested, but there is no dedicated LLVM test. I will remove this from this change, and it can be upstreamed separately.

djtodoro committed:
fixup: Improve the aligment check

wangpc-pp reviewed (commented):
A thought:
The lowering path can be:
1.  Merge two consecutive load/stores into **pseudo** load/store pair instructions no matter what the offsets are.
2. Expand these pseudos in `RISCVExpandPseudo` by checking if their offsets are valid.

In this way, other vendor pair extensions like THead's can also benefit from this pass.

djtodoro committed:
fixup: do not rely on stack objects being 16B aligned

topperc commented:
> A thought: The lowering path can be:
> 
> 1. Merge two consecutive load/stores into **pseudo** load/store pair instructions no matter what the offsets are.
> 2. Expand these pseudos in `RISCVExpandPseudo` by checking if their offsets are valid.
> 
> In this way, other vendor pair extensions like THead's can also benefit from this pass.

Why couldn't we check the offsets in this pass?

wangpc-pp commented:
> > A thought: The lowering path can be:
> > 
> > 1. Merge two consecutive load/stores into **pseudo** load/store pair instructions no matter what the offsets are.
> > 2. Expand these pseudos in `RISCVExpandPseudo` by checking if their offsets are valid.
> > 
> > In this way, other vendor pair extensions like THead's can also benefit from this pass.
> 
> Why couldn't we check the offsets in this pass?

Yes we could. I was thinking that we may make use of PseudoInstExpansion.

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,397 @@
+// Merge two adjacent load/store instructions into a paired instruction
+// (LDP/SDP/SWP/LWP) if the effective address is 16-byte aligned. This function
+// selects the appropriate paired opcode, verifies that the memory operand (or
+// fixed-stack slot) is 16-byte aligned, and checks that the offset is valid. If
```
Is this a requirement of the instruction or just for performance reasons?

djtodoro replied:
Well, for this CPU:

```
Description: An LDP instruction is guaranteed to atomically read 16-byte data if its address is 16-byte aligned. If
the address is 8-byte aligned but not 16-byte aligned, then the data is only guaranteed to be 8-byte atomic. For any
other alignment, there is no atomicity guarantee.
A Load Address Misaligned exception may occur if the virtual address targeted by the LDP instruction is not
aligned to a 16-byte boundary. Performance optimized implementations will provide native hardware support for 8-
byte aligned cases, including cases that cross a page boundary. When $rd1 and $rd2 are the same register, the
value written to the output register is unknown.
```

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,391 @@
+  const MachineMemOperand *MMO = *First->memoperands_begin();
+  Align MMOAlign = MMO->getAlign();
+
+  // Only pair if alignment is exactly 16 bytes.
```
If's LWP/SWP you should be able to use 8 byte alignment right?

djtodoro replied:
Well, yes, it was implemented in original patch, I dropped it in those fixups, wasn't intentional. Thanks.

djtodoro committed:
fixup: fix aligment of lwp/swp

djtodoro commented:
Is this good to go?

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -2750,6 +2750,46 @@ MachineInstr *RISCVInstrInfo::emitLdStWithAddr(MachineInstr &MemI,
+bool RISCVInstrInfo::isPairableLdStInstOpc(unsigned Opc) {
+  switch (Opc) {
+  default:
+    return false;
```
I don't understand the contents of this list. It has more than what Mips supports, but it is not all load and store opcodes. LB, LBU, and LWU are missing.

djtodoro replied:
Well, we meant to put here all instructions for which it is possible to implement pairing, but, it is confusing. I will add MIPS pairable instruction only for now here.

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,397 @@
+  Align MMOAlign = MMO->getAlign();
+
+  // Only pair if alignment is exactly RequiredAlignment bytes.
+  if (MMOAlign != RequiredAlignment)
```
Should this be `MMOAlign < RequiredAlignment`. Being more aligned should be fine.

djtodoro replied:
Yes, that is true

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,397 @@
+
+  int64_t Offset = First->getOperand(2).getImm();
+  if (!isUInt<7>(Offset) ||
+      !isAligned(Align(MMO->getSize().getValue()), Offset))
```
Why is there a separate alignment check here?

djtodoro replied:
not needed

topperc commented on llvm/lib/Target/RISCV/RISCVLoadStoreOptimizer.cpp:
```diff
@@ -0,0 +1,397 @@
+  // Kill flags may become invalid when moving stores for pairing.
+  if (I->getOperand(0).isUse()) {
+    if (!MergeForward) {
+      // Clear kill flags on store if moving upwards.
```
Doesn't the equivalent part of the AArch64 code have a loop here? Why don't we need that loop?

djtodoro committed:
fixup: Improve fix flags and use MIPS pariable instrs only

djtodoro commented:
hey @topperc thanks a lot for you comments!

topperc commented on llvm/test/CodeGen/RISCV/load-store-pair.ll:
```diff
@@ -0,0 +1,326 @@
+; RV64D_NOPAIR-NEXT:    .cfi_def_cfa_offset 0
+; RV64D_NOPAIR-NEXT:    ret
+entry:
+  %arrayidx = getelementptr inbounds i8*, i8** %a, i64 1
```
use opaque pointers. i8* and i8** syntax is deprecated.

djtodoro replied:
It makes sense, done.

topperc commented on llvm/test/CodeGen/RISCV/load-store-pair.ll:
```diff
@@ -0,0 +1,326 @@
+; RUN: llc -mtriple=riscv64 -target-abi=lp64d -mattr=+d -verify-machineinstrs < %s \
+; RUN:   | FileCheck %s -check-prefix=RV64D_NOPAIR
+
+define dso_local void @testi(i8** nocapture noundef readonly %a) local_unnamed_addr #0 {
```
Remove dso_local and local_unnamed_addr and #0

djtodoro committed:
fixup: Port test to opaque pointers

topperc reviewed (approved):
LGTM
